SpatialGDK/Source/SpatialGDKFunctionalTests/Private/SpatialFunctionalTestSoftAssertHandler.cpp

```diff
@@ -0,0 +1,393 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "SpatialFunctionalTestSoftAssertHandler.h"
+#include "Logging/LogMacros.h"
+#include "Misc/AssertionMacros.h"
+#include "SpatialFunctionalTest.h"
+#include "SpatialFunctionalTestFlowController.h"
+#include "SpatialGDKFunctionalTestsPrivate.h"
+#include "VisualLogger/VisualLogger.h"
+
+namespace
+{
+template <typename T>
+bool Compare(const T& A, EComparisonMethod Operator, const T& B)
+{
+	bool bPassed = false;
+	switch (Operator)
+	{
+	case EComparisonMethod::Equal_To:
+		bPassed = A == B;
+		break;
+	case EComparisonMethod::Not_Equal_To:
+		bPassed = A != B;
+		break;
+	case EComparisonMethod::Greater_Than_Or_Equal_To:
+		bPassed = A >= B;
+		break;
+	case EComparisonMethod::Less_Than_Or_Equal_To:
+		bPassed = A <= B;
+		break;
+	case EComparisonMethod::Greater_Than:
+		bPassed = A > B;
+		break;
+	case EComparisonMethod::Less_Than:
+		bPassed = A < B;
+		break;
+	}
+	return bPassed;
+}
+
+FString GetComparisonMethodAsString(EComparisonMethod Operator)
+{
+	switch (Operator)
+	{
+	case EComparisonMethod::Equal_To:
+		return TEXT("==");
+	case EComparisonMethod::Not_Equal_To:
+		return TEXT("!=");
+	case EComparisonMethod::Greater_Than_Or_Equal_To:
+		return TEXT(">=");
+	case EComparisonMethod::Less_Than_Or_Equal_To:
+		return TEXT("<=");
+	case EComparisonMethod::Greater_Than:
+		return TEXT(">");
+	case EComparisonMethod::Less_Than:
+		return TEXT("<");
+	}
+
+	check(false);
+	return FString(); // For compilation.
+}
+
+FString GetTransformAsString(const FTransform& Transform)
+{
+	FVector T = Transform.GetLocation();
+	FRotator R = Transform.Rotator();
+	FVector S = Transform.GetScale3D();
+	return FString::Printf(TEXT("T(%f,%f,%f) | R(%f, %f, %f), S(%f, %f, %f)"), T.X, T.Y, T.Z, R.Pitch, R.Yaw, R.Roll, S.X, S.Y, S.Z);
+}
+} // namespace
+
+SpatialFunctionalTestSoftAssertHandler::SpatialFunctionalTestSoftAssertHandler()
+	: NextOrder(0)
+{
+}
+
+void SpatialFunctionalTestSoftAssertHandler::SoftAssertTrue(bool bCheckTrue, const FString& Msg)
+{
+	GenericSoftAssert(Msg, bCheckTrue, FString());
+}
+
+void SpatialFunctionalTestSoftAssertHandler::SoftAssertFalse(bool bCheckFalse, const FString& Msg)
+{
+	GenericSoftAssert(Msg, !bCheckFalse, FString());
+}
+
+void SpatialFunctionalTestSoftAssertHandler::SoftAssertCompare(int A, EComparisonMethod Operator, int B, const FString& Msg)
+{
+	bool bPassed = Compare(A, Operator, B);
+
+	FString ErrorMsg;
+
+	if (!bPassed)
+	{
+		ErrorMsg = FString::Printf(TEXT("Expected %d %s %d"), A, *GetComparisonMethodAsString(Operator), B);
+	}
+
+	GenericSoftAssert(Msg, bPassed, ErrorMsg);
+}
+
+void SpatialFunctionalTestSoftAssertHandler::SoftAssertCompare(float A, EComparisonMethod Operator, float B, const FString& Msg)
```

Comment:
Maybe we could do some template magic for these...
But I have always been bad at that, so someone else may have a more informed opinion.

Reply:
The reason why I don't use templates is because I create an error message only when it fails, and the error messages unfortunately depend on different parameters. If we want to create that string every call, I can refactor this into a templated kind of function. What do you prefer?